File: 2017/05/05.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	// Part one
	ints := ReadIntLines("input")
	fmt.Println(jump(ints, partOne))

	// Part two
	fmt.Println(jump(ints, partTwo))
}

func ReadIntLines(filename string) []int {
	f, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	s := bufio.NewScanner(f)
	s.Split(bufio.ScanLines)

	var ints []int
	for s.Scan() {
		i, err := strconv.Atoi(s.Text())
		if err != nil {
			panic(err)
		}
		ints = append(ints, i)
	}
	return ints
}

func jump(lines []int, update func([]int, int)) int {
	ints := make([]int, len(lines))
	copy(ints, lines)

	current := 0
	nbJumps := 0
	for current >= 0 && current < len(ints) {
		step := ints[current]
		update(ints, current)

		current += step
		nbJumps++
	}
	return nbJumps
}

func partOne(ints []int, index int) {
	ints[index]++
}

func partTwo(ints []int, index int) {
	if ints[index] >= 3 {
		ints[index]--
	} else {
		ints[index]++
	}
}

File: 2017/05/05.py
import time

def jump(lines, update):
    current, nb_jumps = 0, 0
    while current >= 0 and current < len(lines):
        steps = lines[current]
        update(lines, current)
        # Jump
        current += steps
        nb_jumps += 1
    return nb_jumps

def partOne(lines, current):
    lines[current] += 1

def partTwo(lines, current):
    lines[current] += -1 if lines[current] >= 3 else 1

with open('input', 'r') as f:
    lines = [int(line) for line in f.read().splitlines()]
    # f.readlines() = list(f) = [l for l in f] # Keeps \n at the end of lines

    start_time = time.time()
    print(jump(lines[:], partOne))
    print("1: %s ms" % float((time.time() - start_time)*1000))

    start_time = time.time()
    print(jump(lines[:], partTwo))
    print("2: %s ms" % float((time.time() - start_time)*1000))
